src/solver/tron.jl

```diff
@@ -0,0 +1,414 @@
+# Some parts of this code were adapted from
+# https://github.com/PythonOptimizers/NLP.py/blob/develop/nlp/optimize/tron.py
+
+using LinearOperators, NLPModels
+
+export tron, project!, project_step!, active, breakpoints
+
+"""`tron(nlp)`
+
+A pure Julia implementation of a trust-region solver for bound-constrained
+optimization:
+
+    min f(x)    s.t.    ℓ ≦ x ≦ u
+
+TRON is described in
+
+Chih-Jen Lin and Jorge J. Moré, *Newton's Method for Large Bound-Constrained
+Optimization Problems*, SIAM J. Optim., 9(4), 1100–1127, 1999.
+"""
+function tron(nlp :: AbstractNLPModel;
+              μ₀ :: Real=1e-2,
+              μ₁ :: Real=1.0,
+              σ :: Real=10.0,
+              verbose=false,
+              itmax :: Int=10_000*nlp.meta.nvar,
+              max_cgiter :: Int=nlp.meta.nvar,
+              cgtol = 0.1,
+              fmin :: Real = -1e32,
+              timemax :: Real=60.0,
+              mem :: Integer=5,
+              atol :: Real=1e-8,
+              rtol :: Real=1e-6,
+              fatol :: Real=0.0,
+              frtol :: Real=1e-12,
+              σ₁ :: Real=0.25,
+              σ₂ :: Real=0.5,
+              σ₃ :: Real=4.0,
+              η₀ :: Real=1e-4,
+              η₁ :: Real=0.25,
+              η₂ :: Real=0.75,
+              Δmax :: Real=1e6
```

Comment:
Could the trust-region parameters somehow be associated with a trust-region "object" as in Trunk?